contracts/modules/v2/StakingPool.sol

```diff
@@ -349,71 +378,122 @@ contract StakingPool is ERC20 {
+    return pricePercentage * coverAmount / MAX_PRICE_PERCENTAGE * period / 365 days;
   }
 
-  function getCapacity(uint productId, uint capacityFactor) public view returns (uint) {
-    return 0;
-  }
+  uint public constant SURGE_THRESHOLD = 8e17;
+  uint public constant BASE_SURGE_LOADING = 1e16;
 
-  function calculatePremium(uint pricePercentage, uint coverAmount, uint period) public pure returns (uint) {
-    return pricePercentage * coverAmount / MAX_PRICE_PERCENTAGE * period / 365 days;
+
+  function getActualPriceAndUpdateBasePrice(
+    uint productId,
+    uint amount,
+    uint activeCover,
+    uint capacity,
+    uint initialPrice
+  ) internal returns (uint) {
+
+
+    (uint actualPrice, uint basePrice) = getPrices(productId, amount, activeCover, capacity, initialPrice);
+    // store the last base price
+    lastPrices[productId] = LastPrice(
+      uint96(basePrice),
+      uint32(block.timestamp
+    ));
+
+    return actualPrice;
   }
 
-  function calculatePrice(
+  function getPrices(
+    uint productId,
     uint amount,
-    uint basePrice,
     uint activeCover,
-    uint capacity
-  ) public pure returns (uint) {
+    uint capacity,
+    uint initialPrice
+  ) public view returns  (uint actualPrice, uint basePrice) {
+    uint96 lastPrice = lastPrices[productId].value;
+    basePrice = interpolatePrice(
+      lastPrice != 0 ? lastPrice : initialPrice,
+      targetPrices[productId],
+      lastPrices[productId].lastUpdateTime,
+      block.timestamp
+    );
+
+    // calculate actualPrice using the current basePrice
+    actualPrice = calculatePrice(amount, basePrice, activeCover, capacity);
 
-    return (calculatePriceIntegralAtPoint(
-      basePrice,
-      activeCover + amount,
-      capacity
-    ) -
-    calculatePriceIntegralAtPoint(
-      basePrice,
-      activeCover,
-      capacity
-    )) / amount;
+    // Bump base price by 2% (200 basis points) per 10% (1000 basis points) of capacity used
+    uint priceBump = amount * BASIS_PRECISION / capacity / 1000 *  200;
+    basePrice = uint96(basePrice + priceBump);
   }
 
-  function calculatePriceIntegralAtPoint(
+  function calculatePrice(
+    uint amount,
     uint basePrice,
     uint activeCover,
     uint capacity
   ) public pure returns (uint) {
-    uint actualPrice = basePrice * activeCover;
-    for (uint i = 0; i < PRICE_CURVE_EXPONENT; i++) {
-      actualPrice = actualPrice * activeCover / capacity;
+
+    uint newActiveCoverAmount = amount + activeCover;
+    uint newActiveCoverPercentage = newActiveCoverAmount * 1e18 / capacity;
+
+    if (newActiveCoverPercentage > SURGE_THRESHOLD) {
+      return basePrice;
     }
-    actualPrice = actualPrice / 8 + basePrice * activeCover;
 
-    return actualPrice;
+    uint surgeLoadingPercentage = newActiveCoverPercentage - SURGE_THRESHOLD;
+    uint surgeFraction = surgeLoadingPercentage * capacity / newActiveCoverAmount;
```

Comment:
Let's stick to one naming convention for ratios instead of coming up with new suffixes every time (fraction, percentage, bps, rate) and give more meaningful names to variables to avoid confusion. What would be the difference between surgeFraction and surgeLoadingPercentage?